### PR TITLE
Refactor indices aliases

### PR DESCRIPTION
**Pull request by joegallo**
Please don't actually merge this. I'd rather we use git trickery to update your branch **without** merging this PR. But I did want some easy way to say "look at what I've done here since we last spoke" and this is a half decent way of doing that.

Annoyingly there's some (IMHO) unnecessary noise because I merged master in.

In conclusion, reiterating: please don't actually merge this.

❤️

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 1]
```diff
@@ -784,12 +784,20 @@ public IndexMetadata getIndexSafe(Index index) {
...
+        return aliases();
+    }
+
     public ImmutableOpenMap<String, IndexTemplateMetadata> templates() {
         return this.templates;
     }
 
     public ImmutableOpenMap<String, IndexTemplateMetadata> getTemplates() {
-        return this.templates;
+        return templates();
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:800]: unrelated change?

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 2]
```diff
@@ -1229,6 +1237,28 @@ public Builder indices(ImmutableOpenMap<String, IndexMetadata> indices) {
...
+            indices.add(index); // TODO what if it's a duplicate? maybe use a Set<Index> instead?
+            aliases.put(alias, indices);
+            return this;
+        }
+
+        public Builder removeAlias(String alias, Index index) {
+            previousIndicesLookup = null;
+
+            List<Index> indices = new ArrayList<>(aliases.getOrDefault(alias, List.of()));
+            indices.remove(index); // TODO what if it wasn't in the list already?
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1253]: maybe: 

```java
if (indices.remove(index) == false) {
   return this;
}
```

Which would make this a no-op.

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 3]
```diff
@@ -1229,6 +1237,28 @@ public Builder indices(ImmutableOpenMap<String, IndexMetadata> indices) {
             return this;
         }
 
+        public Builder putAlias(String alias, Index index) {
+            previousIndicesLookup = null;
+
+            List<Index> indices = new ArrayList<>(aliases.getOrDefault(alias, List.of()));
+            indices.add(index); // TODO what if it's a duplicate? maybe use a Set<Index> instead?
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1244]: I think it easier if we make we change the aliases map to contain a Set<Index> instead of if List<Index>.

and also maybe:

```java
if (indices.add(index) == false) {
    return this;
}
```

to make this a no-op in case .

File: server/src/test/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexServiceTests.java [hunk 4]
```diff
@@ -121,6 +121,31 @@ public void testDeleteUnassigned() {
...
+            .numberOfShards(1)
+            .numberOfReplicas(1)
+            .build();
+        ClusterState before = ClusterState.builder(ClusterName.DEFAULT)
+            .metadata(Metadata.builder().put(idxMetadata, false).putAlias(alias, idxMetadata.getIndex()))
+            .routingTable(RoutingTable.builder().addAsNew(idxMetadata).build())
+            .blocks(ClusterBlocks.builder().addBlocks(idxMetadata))
+            .build();
+
+        ClusterState after = service.deleteIndices(before, Set.of(before.metadata().getIndices().get(index).getIndex()));
```
Comment (martijnvg) [inline server/src/test/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexServiceTests.java:140]: maybe check that that before the delete call an alias entry does exist?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexService.java [hunk 5]
```diff
@@ -124,6 +124,13 @@ public ClusterState deleteIndices(ClusterState currentState, Set<Index> indices)
                 metadataBuilder.put(parent.removeBackingIndex(index));
             }
         }
+        // delete associated aliases from the cluster state
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexService.java:127]: Maybe inline this into `Metadata.Builder#remove(...)` method? If an IndexMetadata has aliases then we always need to invoke the `remobeAlias()` method (and we don't need to check whether aliases have changed).

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java (hunk 1 above)
Comment (joegallo) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:800]: Not entirely unrelated, see 7f25f49.

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java (hunk 3 above)
Comment (joegallo) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1244]: ++, I'll make the `List<Index>` --> `Set<Index>` change throughout, and let's see what that gets us.

Comment (joegallo) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1244]: Oh, actually, I just remembered -- https://github.com/elastic/elasticsearch/blob/fed20aa74b48b2220870c782027a69de305bf56c/server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java#L1147-L1150 makes me think there's existing semantics around the order of the aliases for an index, might we want to preserve something that like here? That is, is there value in having List semantics rather than Set semantics or no?

Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1244]: I don't think that is the case, once an index and its aliases are created.
